[InstCombine] Optimize (select %x, op(%x), 0) to op(%x) for operations where op(0) == 0

bababuck opened the pull request:
Currently this optimization only occurs for `mul`, but this generalizes that for any operation that has a fixed point of `0`.

There is similar logic within `EarlyCSE` pass, but that is stricter in terms of `poison` propagation so will not optimize for many operations.

Alive2 Proofs:
`and`:
https://alive2.llvm.org/ce/z/RraasX ; base-case
https://alive2.llvm.org/ce/z/gzfFTX ; commuted-case
https://alive2.llvm.org/ce/z/63XaoX ; compare against undef
https://alive2.llvm.org/ce/z/MVRVNd ; select undef
https://alive2.llvm.org/ce/z/2bsoYG ; vector
https://alive2.llvm.org/ce/z/xByeX- ; vector compare against undef
https://alive2.llvm.org/ce/z/zNdzmZ ; vector select undef

`fshl`:
https://alive2.llvm.org/ce/z/U3_PG3 ; base-case
https://alive2.llvm.org/ce/z/BWCnxT ; compare against undef
https://alive2.llvm.org/ce/z/8HGAE_ ; select undef
; vector times out

`fshr`:
https://alive2.llvm.org/ce/z/o6F47G ; base-case
https://alive2.llvm.org/ce/z/fVnBXy ; compare against undef
https://alive2.llvm.org/ce/z/suymYJ ; select undef
; vector times out

`umin`:
https://alive2.llvm.org/ce/z/GGMqf6 ; base-case
https://alive2.llvm.org/ce/z/6cx5-k ; commuted-case
https://alive2.llvm.org/ce/z/W5d9tz ; compare against undef
https://alive2.llvm.org/ce/z/nKbaUn ; select undef
https://alive2.llvm.org/ce/z/gxEGqc ; vector
https://alive2.llvm.org/ce/z/_SDpi_ ; vector compare against undef

`sdiv`:
https://alive2.llvm.org/ce/z/5XGs3q

`srem`:
https://alive2.llvm.org/ce/z/vXAnQM

`udiv`:
https://alive2.llvm.org/ce/z/e6_8Ug

`urem`:
https://alive2.llvm.org/ce/z/VmM2SL

`shl`:
https://alive2.llvm.org/ce/z/aCZr3u ; Argument with range
https://alive2.llvm.org/ce/z/YgDy8C ; Instruction with known bits
https://alive2.llvm.org/ce/z/6pIxR6 ; Constant

`lshr`:
https://alive2.llvm.org/ce/z/WCCBej

`ashr:
https://alive2.llvm.org/ce/z/egV4TR

Commit:
Add tests for cases with (select %x, op(%x), 0) for operations where op(0) == 0

These cases can be optimized to just op(%x).

Commit:
[InstCombine] Optimize (select %x, op(%x), 0) to op(%x) for operations where op(0) == 0

Have to freeze the any other operands to prevent poisons from leaking. Re-uses
flow from `mul` specific version of this within the InstCombie pass.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -913,11 +915,28 @@ static Instruction *foldSelectZeroOrMul(SelectInst &SI, InstCombinerImpl &IC) {
-  return IC.replaceInstUsesWith(SI, FalseValI);
+  if (match(FalseVal, m_c_Mul(m_Specific(X), m_Value(Y))) ||
+      match(FalseVal, m_c_And(m_Specific(X), m_Value(Y))) ||
+      match(FalseVal, m_Shl(m_Specific(X), m_Value(Y))) ||
```
It doesn't work for shift operators. `shl 0, freeze(poison)` can be poison when shamt is invalid.

bababuck replied:
Thanks, I will remove `shl`, `ashr` and `lshr`.

From what I can tell, `fshl` and `fshr` intrinsics do not have this issue so they can remain. Is that correct?

bababuck replied:
If I wanted to perform this transformation on the shift operators (`shl`, `ashr`, `lshr`), should that get done during target specific lowering (i.e. RISCV `sll` instruction uses modulo bitwidth for shamt)?

dtcxzyw replied:
> If I wanted to perform this transformation on the shift operators (shl, ashr, lshr), should that get done during target specific lowering (i.e. RISCV sll instruction uses modulo bitwidth for shamt)?

You can check if `Y` is never poison with `isGuaranteedNotTobePoison`.

and/mul: valid with freeze(Y) if `Y` may be poison
shl/lshr/ashr: valid if `Y` cannot be poison
udiv/sdiv/urem/srem: always valid without freeze

bababuck replied:
https://alive2.llvm.org/ce/z/GEJeqY

bababuck replied:
I am going to leave the shift's out from this MR for the following reason. For shift operations, if the inputs are guaranteed not to be poison, whether or not the operation `isGuaranteedNotToBePoison` boils down to the `shamt`, and a function call to `static bool shiftAmountKnownInRange(const Value *ShiftAmount)`, which checks if the `shamt` is a Constant smaller than the bit-width.

However, there is an existing transformation in `InstCombine` that handles the cases for which the `shamt` is a constant, so it seems redundant to include it here.

A separate note, if we could get `static bool shiftAmountKnownInRange(const Value *ShiftAmount)` access to information about Value's ranges (i.e. range(i64 0 10)) it would be effective in more situations.

dtcxzyw replied:
I think we can start with constant (splat) shamts (i.e., use `m_APInt(C)` and `C->ult(BW)`).

> A separate note, if we could get static bool shiftAmountKnownInRange(const Value *ShiftAmount) access to information about Value's ranges (i.e. range(i64 0 10)) it would be effective in more situations.

If you want to handle variable shamts, use `computeKnownBits().getMaxValue().ult(BW)`.

nikic replied:
Can we please keep the shift case out of this patch? It requires changing generic ValueTracking APIs, and we need to evaluate the impact of that separately.

bababuck replied:
I'll remove it here and then create a separate MR for that once this gets merged.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -900,9 +904,7 @@ static Instruction *foldSelectZeroOrMul(SelectInst &SI, InstCombinerImpl &IC) {
-  if (TrueValC == nullptr ||
-      !match(FalseVal, m_c_Mul(m_Specific(X), m_Value(Y))) ||
-      !isa<Instruction>(FalseVal))
+  if (TrueValC == nullptr || !isa<Instruction>(FalseVal))
```
Crash reproducer:
```
; bin/opt -passes=instcombine test.ll -S
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu"

define <2 x i64> @php_url_encode_impl(i32 %0, ptr %p) {
  %2 = load <2 x i64>, ptr %p, align 16
  %.not = icmp eq i32 %0, 0
  %spec.select = select i1 %.not, <2 x i64> zeroinitializer, <2 x i64> %2
  ret <2 x i64> %spec.select
}
```
```
opt: /home/dtcxzyw/WorkSpace/Projects/compilers/llvm-project/llvm/lib/IR/Constants.cpp:823: static llvm::Constant* llvm::Constant::mergeUndefsWith(llvm::Constant*, llvm::Constant*): Assertion `isa<FixedVectorType>(Other->getType()) && cast<FixedVectorType>(Other->getType())->getNumElements() == NumElts && "Type mismatch"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: bin/opt -passes=instcombine reduced.ll -S
1.      Running pass "function(instcombine<max-iterations=1;verify-fixpoint>)" on module "reduced.ll"
2.      Running pass "instcombine<max-iterations=1;verify-fixpoint>" on function "php_url_encode_impl"
 #0 0x0000727c5de283e2 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.21.0git+0x2283e2)
 #1 0x0000727c5de249ef llvm::sys::RunSignalHandlers() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMSupport.so.21.0git+0x2249ef)
 #2 0x0000727c5de24b34 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
 #3 0x0000727c5d845330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #4 0x0000727c5d89eb2c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x0000727c5d89eb2c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x0000727c5d89eb2c pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x0000727c5d84527e raise ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x0000727c5d8288ff abort ./stdlib/abort.c:81:7
 #9 0x0000727c5d82881b _nl_load_domain ./intl/loadmsgcat.c:1177:9
#10 0x0000727c5d83b517 (/lib/x86_64-linux-gnu/libc.so.6+0x3b517)
#11 0x0000727c54587344 llvm::Constant::mergeUndefsWith(llvm::Constant*, llvm::Constant*) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.21.0git+0x187344)
#12 0x0000727c55969950 foldSelectZeroOrFixedOp(llvm::SelectInst&, llvm::InstCombinerImpl&) InstCombineSelect.cpp:0:0
#13 0x0000727c559709d0 llvm::InstCombinerImpl::visitSelectInst(llvm::SelectInst&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.21.0git+0x16f9d0)
#14 0x0000727c55869c18 llvm::InstCombinerImpl::run() (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.21.0git+0x68c18)
#15 0x0000727c5586ad7f combineInstructionsOverFunction(llvm::Function&, llvm::InstructionWorklist&, llvm::AAResults*, llvm::AssumptionCache&, llvm::TargetLibraryInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::OptimizationRemarkEmitter&, llvm::BlockFrequencyInfo*, llvm::BranchProbabilityInfo*, llvm::ProfileSummaryInfo*, llvm::InstCombineOptions const&) InstructionCombining.cpp:0:0
#16 0x0000727c5586bde4 llvm::InstCombinePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMInstCombine.so.21.0git+0x6ade4)
#17 0x0000727c57daa635 llvm::detail::PassModel<llvm::Function, llvm::InstCombinePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libPolly.so.21.0git+0x1aa635)
#18 0x0000727c54733a15 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.21.0git+0x333a15)
#19 0x0000727c5c8da9b5 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.21.0git+0xda9b5)
#20 0x0000727c54733e12 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.21.0git+0x333e12)
#21 0x0000727c5c8db375 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMX86CodeGen.so.21.0git+0xdb375)
#22 0x0000727c54734fed llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/../lib/libLLVMCore.so.21.0git+0x334fed)
#23 0x0000727c5df8e309 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.21.0git+0x2d309)
#24 0x0000727c5df994c2 optMain (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/bin/../lib/libLLVMOptDriver.so.21.0git+0x384c2)
#25 0x0000727c5d82a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#26 0x0000727c5d82a28b call_init ./csu/../csu/libc-start.c:128:20
#27 0x0000727c5d82a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#28 0x000059b9a636b095 _start (bin/opt+0x1095)
Aborted (core dumped)
```

bababuck replied:
Thanks.

The issue was that I moved the pattern match to the end of the function. However, one side effect of the pattern match is that it guarantees that the types of TrueV and the Conditional constant Match which is assumed to be true by:
```
auto *MergedC = Constant::mergeUndefsWith(TrueValC, ZeroC);
```

Reordering to perform the pattern matching where it was before this MR solved this issue.

nikic reviewed (commented):
Can you please add alive2 proofs to the PR description?

bababuck committed:
Reorder so that matching occurs before any data extraction

Fixes bug in optimizing:
```
define <2 x i64> @php_url_encode_impl(i32 %0, ptr %p) {
  %2 = load <2 x i64>, ptr %p, align 16
  %.not = icmp eq i32 %0, 0
  %spec.select = select i1 %.not, <2 x i64> zeroinitializer, <2 x i64> %2
  ret <2 x i64> %spec.select
}
```

One side effect of the matching is that it garuntees that the types of the TrueV and the Conditional constant match which is assumed by the later code.

bababuck committed:
Use m_c_Intrinsic<Intrinisic::umin> instead of m_c_UMin

m_c_UMin will also match with an icmp/select pattern which isn't
desired here.

bababuck committed:
Don't match with `shl`, `ashr`, and `ashl` because if the `shfamt` is too large we can create a `poison` value

bababuck committed:
Don't match with `udiv` and `sdiv` because if the denominator is 0 we can create `undef` value

bababuck commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -913,11 +915,28 @@ static Instruction *foldSelectZeroOrMul(SelectInst &SI, InstCombinerImpl &IC) {
+      match(FalseVal, m_FShl(m_Specific(X), m_Specific(X), m_Value(Y))) ||
+      match(FalseVal, m_FShr(m_Specific(X), m_Specific(X), m_Value(Y))) ||
+      match(FalseVal, m_SDiv(m_Specific(X), m_Value(Y))) ||
+      match(FalseVal, m_UDiv(m_Specific(X), m_Value(Y))) ||
```
For more context, Alive2 doesn't flag the `[s/u]div` transformation.

dtcxzyw replied:
> Should `sdiv` and `udiv` should be removed as well since they can create `undef` in the case of the denominator being `0`? My understanding is that `0 / freeze(y)` would be more poisonous than `0` in that case when `freeze(y)` is `0`.

It is safe, even without inserting freeze. If `Y` is poison, it means `div/rem X, Y` triggers immediate UB before the select. Therefore, `Y` cannot be poison.

bababuck replied:
Thanks for the explanation, I appreciate it!

bababuck committed:
Add back in SDiv/UDiv/URem/SRem, but don't freeze them

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -7318,6 +7318,25 @@ bool llvm::isOverflowIntrinsicNoWrap(const WithOverflowInst *WO,
+  if (A) {
+    auto Range = A->getRange();
+    if (Range && Range->getUnsignedMax().ult(
+                     ShiftAmount->getType()->getIntegerBitWidth())) {
```
This should use getScalarSizeInBits(), otherwise it will crash for vector shifts. Same below.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -900,10 +904,25 @@ static Instruction *foldSelectZeroOrMul(SelectInst &SI, InstCombinerImpl &IC) {
+      match(FalseVal,
+            m_c_Intrinsic<Intrinsic::umin>(m_Specific(X), m_Value(Y)))) {
+    FreezeY = true;
+  } else if (match(FalseVal, m_SDiv(m_Specific(X), m_Value(Y))) ||
+             match(FalseVal, m_UDiv(m_Specific(X), m_Value(Y))) ||
+             match(FalseVal, m_SRem(m_Specific(X), m_Value(Y))) ||
+             match(FalseVal, m_URem(m_Specific(X), m_Value(Y)))) {
```
```suggestion
  } else if (match(FalseVal, m_IDiv(m_Specific(X), m_Value(Y))) ||
             match(FalseVal, m_IRem(m_Specific(X), m_Value(Y)))) {
```

bababuck committed:
Match to m_I[Div/Rem] rather than the individual signed and unsigned versions

No functional change, reduces code size.

Co-authored-by: Yingwei Zheng <dtcxzyw@qq.com>

nikic commented on llvm/test/Transforms/InstCombine/select-fixed-zero.ll:
```diff
@@ -0,0 +1,195 @@
+  %mul = mul i64 %a, %b
+  %select = select i1 %cond, i64 0, i64 %mul
+  ret i64 %select
+}
```
Test commuted operands for mul/and?

nikic reviewed (approved):
LGTM

bababuck committed:
Add tests for commuted versions of AND and MUL

bababuck commented:
Rebased to latest `main` added test for commuted cases of `and` and `mul`
